specs/core/1_beacon-chain-misc.md

```diff
@@ -0,0 +1,218 @@
+    """
+    Computes the generalized index of the state root of slot `frm` based on the state root of slot `to`.
+    Relies on the `history_acc` in the `ShardState`, where `history_acc[i]` maintains the most recent 2**i'th
+    slot state. Works by tracing a `log(later-earlier)` step path from `later` to `earlier` through intermediate
+    blocks at the next available multiples of descending powers of two.
+    """
+    o = GeneralizedIndex(1)
+    for i in range(63, -1, -1):
+        if (later - 1) & 2**i > (earlier - 1) & 2**i:
+            later = later - ((later - 1) % 2**i) - 1
+            o = concat_generalized_indices(o, get_generalized_index(ShardState, 'history_acc', i))
+    return o
+```
+
+#### `get_generalized_index_of_crosslink_header`
+
+```python
+def get_generalized_index_of_crosslink_header(index: int) -> GeneralizedIndex:
+    """
+    Gets the generalized index for the root of the index'th header in a crosslink.
+    """
+    MAX_CROSSLINK_SIZE = SHARD_BLOCK_SIZE_LIMIT * SHARD_SLOTS_PER_BEACON_SLOT * SLOTS_PER_EPOCH * MAX_EPOCHS_PER_CROSSLINK
+    assert MAX_CROSSLINK_SIZE == get_previous_power_of_2(MAX_CROSSLINK_SIZE)
+    return GeneralizedIndex(MAX_CROSSLINK_SIZE // SHARD_HEADER_SIZE + index)
+```
+
+#### `process_shard_receipt`
+
+```python
+def process_shard_receipt(state: BeaconState, receipt_proof: ShardReceiptProof):
+    """
+    Processes a ShardReceipt object.
+    """
+    receipt_slot = state.next_shard_receipt_period[receipt_proof.shard] * SLOTS_PER_EPOCH * EPOCHS_PER_SHARD_PERIOD
+    first_slot_in_last_crosslink = state.current_crosslinks[receipt_proof.shard].start_epoch * SLOTS_PER_EPOCH
+    gindex = concat_generalized_indices(
+        get_generalized_index_of_crosslink_header(0),
+        get_generalized_index(ShardBlockHeader, 'state_root'),
+        compute_historical_state_generalized_index(receipt_slot, first_slot_in_last_crosslink),
+        get_generalized_index(ShardState, 'receipt_root')
+    )
+    assert verify_merkle_proof(
+        leaf=hash_tree_root(receipt_proof.receipt),
+        proof=receipt_proof.proof,
+        index=gindex,
+        root=state.current_crosslinks[shard].data_root
+    )
+    for delta in receipt_proof.receipt:
+        if get_current_epoch(state) < state.validators[delta.index].withdrawable_epoch:
+            increase_balance(state, delta.index, state.validators[delta.index].effective_balance * delta.reward_coefficient // REWARD_COEFFICIENT_BASE)
+            decrease_balance(state, delta.index, delta.block_fee)
+    state.next_shard_receipt_period[receipt_proof.shard] += 1
+    increase_balance(state, get_beacon_proposer_index(state), MICRO_REWARD)
+```
+
+## Changes
+
+### Persistent committees
+
+Add to the beacon state the following fields:
+
+```python
+# begin insert @persistent_committee_fields
+    previous_persistent_committee_root: Hash
+    current_persistent_committee_root: Hash
+    next_persistent_committee_root: Hash
```

Comment:
Why do we need three persistent committee roots (as opposed to just one)?

Reply:
As discussed, let's turn this into a historical vector that updates at each period